pkg/autodiscovery/providers/prometheus_common.go

```diff
@@ -0,0 +1,226 @@
+func (pc *PrometheusCheck) init() error {
+	pc.initInstances()
+	return pc.initAD()
+}
+
+// initInstances defaults the Instances field in PrometheusCheck
+func (pc *PrometheusCheck) initInstances() {
+	if len(pc.Instances) == 0 {
+		// Put a default config
+		pc.Instances = append(pc.Instances, &OpenmetricsInstance{
+			Metrics:   openmetricsDefaultMetrics,
+			Namespace: openmetricsDefaultNS,
+		})
+		return
+	}
+
+	for _, instance := range pc.Instances {
+		// Default the required config values if not set
+		if len(instance.Metrics) == 0 {
+			instance.Metrics = openmetricsDefaultMetrics
+		}
+	}
+}
+
+// initAD defaults the AD field in PrometheusCheck
+// It also prepares the regex to match the containers by name
+func (pc *PrometheusCheck) initAD() error {
+	if pc.AD == nil {
+		pc.AD = &ADConfig{}
+	}
+
+	pc.AD.defaultAD()
+	return pc.AD.setContainersRegex()
+}
+
+// defaultAD defaults the values of the autodiscovery structure
+func (ad *ADConfig) defaultAD() {
+	if ad.ExcludeAutoconf == nil {
+		// TODO: Implement OOTB autoconf exclusion
+		ad.ExcludeAutoconf = boolPointer(true)
+	}
+
+	if ad.KubeContainerNames == nil {
+		ad.KubeContainerNames = []string{}
+	}
+
+	if ad.KubeAnnotations == nil {
+		ad.KubeAnnotations = &InclExcl{
+			Excl: map[string]string{prometheusScrapeAnnotation: "false"},
+			Incl: map[string]string{prometheusScrapeAnnotation: "true"},
+		}
+		return
+	}
+
+	if ad.KubeAnnotations.Excl == nil {
+		ad.KubeAnnotations.Excl = map[string]string{prometheusScrapeAnnotation: "false"}
+	}
+
+	if ad.KubeAnnotations.Incl == nil {
+		ad.KubeAnnotations.Incl = map[string]string{prometheusScrapeAnnotation: "true"}
+	}
+}
+
+// setContainersRegex precompiles the regex to match the container names for autodiscovery
+// returns an error if the container names cannot be converted to a valid regex
+func (ad *ADConfig) setContainersRegex() error {
+	ad.containersRe = nil
+	if len(ad.KubeContainerNames) == 0 {
+		return nil
+	}
+
+	regexString := strings.Join(ad.KubeContainerNames, "|")
+	re, err := regexp.Compile(regexString)
+	if err != nil {
+		return fmt.Errorf("Invalid container names - regex: '%s': %v", regexString, err)
+	}
+
+	ad.containersRe = re
+	return nil
+}
+
+// matchContainer returns whether a container name matches the 'kubernetes_container_names' configuration
+func (ad *ADConfig) matchContainer(name string) bool {
+	if ad.containersRe == nil {
+		return true
+	}
+	return ad.containersRe.MatchString(name)
+}
+
+// buildURL returns the 'prometheus_url' based on the default values
+// and the prometheus path and port annotations
+func buildURL(annotations map[string]string) string {
+	port := openmetricsDefaultPort
+	if portFromAnnotation, found := annotations[prometheusPortAnnotation]; found {
+		port = portFromAnnotation
+	}
+
+	path := openmetricsDefaultPath
+	if pathFromAnnotation, found := annotations[prometheusPathAnnotation]; found {
+		path = pathFromAnnotation
+	}
+
+	return openmetricsURLPrefix + port + path
```

Comment:
nit: I think it might be safer to do something like: 
```
u, err := url.Parse(fmt.Sprintf("%s:%s", openmetricsURLPrefix, port)
u.Path = path.Join(u.Path, path)
return u.String()
```

Feel free to ignore.

Reply:
In fact the generated URL can contains auto-discovery variables like `%%host%%` or `%%port%%` that makes it invalid for the `url.Parse` function. So we can't use `url.Parse()`